proofreading section Profiling with MAP

pletzer opened the pull request:
Some of the changes:
* title
* objectives
* pictures
* removed summary
* added exercises
originally we thought about using the openmp test case. To run the case for significant amount of time, we had to increase the domain. However, this is no longer necessary for the mpi test case so removed the -nx 256 -ny 256 -nc 256 options. Also, there is no need to compute the check sums since the code is not modified.

pletzer committed:
- removed the -nx -ny and -nc options...

pletzer committed:
proofreading MAP profiling

chrisdjscott commented on _lessons/python-scatter/01-profiling.md:
```diff
@@ -214,13 +212,10 @@ at the page linked above.
-In contrast to cProfile, MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran and F90, as well as Python codes.
-It can be used without code modification.
-MAP can be launched with a GUI and without. The GUI allows the user to navigate through the code and enables them to focus on specific source lines. The "Express Launch", without a GUI, enables easy usage of existing submission scripts and workflows. For more details see the [ARM MAP manual](https://developer.arm.com/docs/101136/latest/map)). On the NeSI system you can start it by loading `module ...
+Another useful profiler provided on the NeSI system is [ARM MAP](https://www.arm.com/products/development-tools/server-and-hpc/forge/map) (previous Allinea MAP) profiler, which is part of the `forge` module (as well as the parallel debugger DDT). In contrast to cProfile, MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran and F90, as well as  ...
```
Links to other lessons should be relative links to the markdown file - it should get updated when deploying to github pages.

```suggestion
Another useful profiler provided on the NeSI system is [ARM MAP](https://www.arm.com/products/development-tools/server-and-hpc/forge/map) profiler, which is part of the `forge` module (along with a parallel debugger, DDT). In contrast to cProfile, MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran and F90, as well as Python codes. More about MAP can be found [here](06-profiling_with_MAP.md).
```

chrisdjscott commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -1,77 +1,79 @@
 Parallel profiling tools provide information about how much time is spent in different parts of your code by different threads/processes.
-There are 2 major ways to collect profiling data: *sampling* and *tracing*. Sampling is a statistical method. During run time the so called *stack trace* is analysed regularly. In contast to that tracing measures actual run time of instrumented parts with a well defined start and end. The data from all threads/processes are gathered and analysed. Thus we can not only see which part of the code is ...
+There are two major ways to collect profiling data: 
+ * *sampling*: data are statistically harvested during run time to determine which parts of the code are most time consuming. This is useful for identiying the hotspots in a code (see [profiling](https://nesi.github.io/perf-training/python-scatter/profiling)).
```
Relative link again:

```suggestion
 * *sampling*: data are statistically harvested during run time to determine which parts of the code are most time consuming. This is useful for identifying the hotspots in a code (see [profiling](01-profiling.md)).
```

chrisdjscott commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -1,77 +1,79 @@
-* representative, need to cover all features used during production
-* short in run time for a fast turnaround, but long enough to be representative of the actual calculations you are going to run
-* scaling effects can be investigated by comparing profiles small and big jobs
+Choose your test case carefully, making sure:
```
```suggestion
Choose your profiling case carefully, making sure:
```

chrisdjscott commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -1,77 +1,79 @@
-MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran and F90, as well as Python codes. It can be used without code modification.
-MAP can be launched with a GUI and without. The GUI allows the user to navigate through the code and enables them to focus on specific source lines. The "Express Launch", without GUI enables easy usage of existing queue scripts and workflows.
+MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran, as well as Python code. It can be used without code modification.
+MAP can be launched with a graphical user interface and without. The graphical user interface allows the user to navigate through the code and focus on specific source lines. The "Express Launch", without graphical user interface, makes it ieasy to submit job scripts and workflows.
```
```suggestion
MAP can be launched with a graphical user interface and without. The graphical user interface allows the user to navigate through the code and focus on specific source lines. The "Express Launch", without the graphical user interface, makes it easy to submit job scripts and workflows.
```

chrisdjscott commented on _lessons/python-scatter/06-profiling_with_MAP.md:
````diff
@@ -1,77 +1,79 @@
+We'll use the *scatter.py* code in directory `mpi` of the *solutions* branch. Start by
 
 ```
+git --fetch all
````
I think you mean:

```suggestion
git fetch --all
```

chrisdjscott commented on _lessons/python-scatter/06-profiling_with_MAP.md:
````diff
@@ -1,77 +1,79 @@
 
 ```
+git --fetch all
+git checkout solutions
````
They would probably need to commit or stash any local changes first but maybe don't need to say that here?

pletzer replied:
Probably not.... They're likely will not have modified the solutions branch

chrisdjscott commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -102,7 +104,7 @@ As an example, "CPU instructions" present the usage of different instruction set
-
-You should now be able to profile parallel programs using MAP to help understand where time is being and where we should start to optimise.
+ * profile the scatter code under the openmp directory using 4 OpenMP threads. Increase the problem size by passing options `-nx 256 -ny 256 -nc 256` to *scatter.py* to make the test run longer.
+ * what is the amount of time spent in computeScatteredWave for the above test case?
```
```suggestion
 * what is the amount of time spent in `computeScatteredWave` for the above test case?
```

mand35 commented on _lessons/python-scatter/01-profiling.md:
```diff
@@ -33,23 +33,21 @@ cd original
-Run the following command to profile the code:
+**Note:** The timing information should not be taken as absolute values, since the profiling itself could extend the run time.
 
+Replace `python scatter.py` with 
```
I would more prefer to write: 
The command line options `-m cProfile -o output.pstats` call the *cProfile* library and writes output the profiling data into "output.pstat". Thus in the present case we call:

pletzer replied:
Thanks. Added:
"Options `-m cProfile -o output.pstats` call the cProfile library and save the profiling data in file *output.pstat*."

chrisdjscott commented on _lessons/python-scatter/06-profiling_with_MAP.md:
````diff
@@ -1,77 +1,79 @@
 ```
 module load forge
-map --profile srun python scatter.py -c -nx 256 -ny 256 -nc 256
+map --profile srun --tasks=8 python scatter.py
````
I don't think the `--tasks=8` is required here? You would set the required number of tasks in an `#SBATCH` line at the top of the submit script and it should be inherited by `srun` - no need to duplicate.

chrisdjscott reviewed (changes requested):
Looks good, I like having some exercises in there. Just some minor changes/suggestions.

mand35 commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -1,77 +1,79 @@
-There are 2 major ways to collect profiling data: *sampling* and *tracing*. Sampling is a statistical method. During run time the so called *stack trace* is analysed regularly. In contast to that tracing measures actual run time of instrumented parts with a well defined start and end. The data from all threads/processes are gathered and analysed. Thus we can not only see which part of the code is ...
+There are two major ways to collect profiling data: 
+ * *sampling*: data are statistically harvested during run time to determine which parts of the code are most time consuming. This is useful for identiying the hotspots in a code (see [profiling](https://nesi.github.io/perf-training/python-scatter/profiling)).
+ * *tracing*: records the activity of instrumented parts of the code in time. This is useful to determine what each thread and process is doing at a particaular point in time. Tracing can reveal load balancing issues or the communication patterns of an MPI program. 
```
thanks for visually highlighting.
Unfortunately it was not clear enough. Tracing measures run time, while sampling "just" statistically analyse call stack snapshots. 
Both have in common that they collect data from all processes ...

pletzer replied:
I was trying to use a simple sentence to express the usefulness of sampling. I'm not sure the users understand what a call stack is. Suffices to say that sapling provides information about where most of the time is spent in a program...

mand35 replied:
Sorry, that additional comment was just for you. 
The point is, that both methods determine what each thread and processes are doing...

pletzer replied:
discussed with Mandes and Wolfgang. Decided we would take out this section

mand35 commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -1,77 +1,79 @@
-* scaling effects can be investigated by comparing profiles small and big jobs
+Choose your test case carefully, making sure:
+* it runs in a short time. Due to a possible large overhead from MAP (especially for tracing experiments), the program could run significantly slower than normal.
+* ensure the test case is representative of a production run
```
Do not miss that the run time should not be too short esp. for sampling. Otherwise, we end up with just a few samples and these will not lead to representative analysis.

pletzer replied:
Thanks @mand35, added "* but the run time should not be too short as this could affect the accuracy of the sampling statistics"

mand35 commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -1,77 +1,79 @@
-MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran and F90, as well as Python codes. It can be used without code modification.
-MAP can be launched with a GUI and without. The GUI allows the user to navigate through the code and enables them to focus on specific source lines. The "Express Launch", without GUI enables easy usage of existing queue scripts and workflows.
+MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran, as well as Python code. It can be used without code modification.
+MAP can be launched with a graphical user interface and without. The graphical user interface allows the user to navigate through the code and focus on specific source lines. The "Express Launch", without graphical user interface, makes it ieasy to submit job scripts and workflows.
```
I would first "graphical user interface (GUI)" and then stay with GUI. 
Furthermore, there is a spelling mistake with ieasy. But actually, we want to prevent using such terms. I would say: "The 'Express Launch', without GUI, allows to maintain workflow, since it just requires a modification of the execution line (*srun* command)."

pletzer replied:
OK thanks.

mand35 commented on _lessons/python-scatter/06-profiling_with_MAP.md:
```diff
@@ -1,77 +1,79 @@
-To use ARM MAP we need to load `module load forge` in our batch script and add `map --profile` in front of the parallel run statements. Thus, `srun` and its options, as well as our executable and its arguments are passed to MAP.
-Thus in the script we will have something like:
-
+To use MAP we need to load the *forge* module in our batch script and add `map --profile` in front of the parallel run statements. For example:
```
To be consistent I would also here write "To use MAP we need to load `module load forge` in our batch script ..."

pletzer replied:
Thanks @mand35. The module load command is written below - no need to duplicate information.

mand35 reviewed (changes requested):
Thanks @pletzer, this looks good, but I have some comments...

chrisdjscott committed:
Update _lessons/python-scatter/06-profiling_with_MAP.md

Co-Authored-By: pletzer <alexander@gokliya.net>

chrisdjscott committed:
Update _lessons/python-scatter/06-profiling_with_MAP.md

Co-Authored-By: pletzer <alexander@gokliya.net>

chrisdjscott committed:
Update _lessons/python-scatter/01-profiling.md

Co-Authored-By: pletzer <alexander@gokliya.net>

chrisdjscott committed:
Update _lessons/python-scatter/06-profiling_with_MAP.md

Co-Authored-By: pletzer <alexander@gokliya.net>

pletzer committed:
going through the comments of PR 53

pletzer committed:
Merge branch 'map' of https://github.com/pletzer/perf-training into map

chrisdjscott committed:
Update _lessons/python-scatter/06-profiling_with_MAP.md

Co-Authored-By: pletzer <alexander@gokliya.net>

pletzer committed:
Merge branch 'map' of https://github.com/pletzer/perf-training into map

pletzer committed:
going through the suggestions

pletzer committed:
back to using GUI acronym

pletzer committed:
+ blurp about cProfile option

pletzer committed:
adding back ticks

pletzer committed:
removed the second explanation of -m cProfile

pletzer committed:
fixing comment

pletzer committed:
Merge branch 'gh-pages' of https://github.com/nesi/perf-training into map

pletzer committed:
~ after discussion with Mandes and Wolfgang resolved to remove the
section describing sampling and tracing as it was not clear whether MAP
was able to do tracing.
